### PR TITLE
SONARPY-2202: Migrate S5727 ComparisonToNoneCheck to the new type model

### PR DESCRIPTION
**Pull request by Seppli11**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java [hunk 1]
```diff
@@ -227,5 +232,40 @@ private static Set<PythonType> collectTypes(PythonType type) {
     }
   }
 
+  record IsIdentityComparableWith(PythonType expectedType) implements TypePredicate {
+    public IsIdentityComparableWith {
+      expectedType = expectedType.unwrappedType();
+    }
 
+    @Override
+    public TriBool test(PythonType otherMaybeWrappedType) {
```
Comment (Seppli11) [inline python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java:241]: I'm not entirely sure that the behaviour of this method and `InferredType.isIdentityComparableWith` is the same.

My understanding of this check is that it checks if two types are the same, or, if one is a union, one of its candidates is the same as the other type. Additionally, if one of the two types is unknown, the result is always `TriBool.UNKNOWN`.

File: python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeCheckerBuilderTest.java
Comment (Seppli11) [inline python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeCheckerBuilderTest.java:1]: was moved to types/v2

**Review by ghislainpiot — CHANGES_REQUESTED** [review]
The utility methods seem to be handling a lot of edge cases that should never happen in the first place.
I think it would be good to create tickets to investigate (and eliminate) the creation of those impossible types.

I didn't review the TypeChecker changes as discussed, I don't think handling the identify case is worth the added complexity on the main predicate.
